### PR TITLE
Support url form encoded bodies

### PR DESCRIPTION
**Pull request by bfrearson**
### Motivation

[Issue 182](https://github.com/apple/swift-openapi-generator/issues/182) Supporting runtime update for [add URL form encoder & decoder](https://github.com/apple/swift-openapi-generator/pull/283)

### Modifications

Add converter methods for Codable <--> URLEncodedForm.

**Note:** Trying to encode a primitive collection (e.g. an array of `String`) form a Codable type fails because the URIEncoder rejects nested object types. I didn't make any changes to the Encoder here, as I think there may be a larger discussion needed about how to approach this. It would be beneficial to support this, as there may be a form which contains multiple values for the same key.

Decoding an exploded key from to a Codable type containing an array works already, but I've not included the test in this PR, as that should be included in the above.

### Result

Converter methods for URLEncodedForms are now available for use in generated content.

### Test Plan

Update tests for encoders and decoders.

### CODE REVIEW

**Comment by czechboy0** [comment]
> Note: Trying to encode a primitive collection (e.g. an array of String) form a Codable type fails because the URIEncoder rejects nested object types. I didn't make any changes to the Encoder here, as I think there may be a larger discussion needed about how to approach this. It would be beneficial to support this, as there may be a form which contains multiple values for the same key.

@bfrearson, please file a separate issue about that if it's specific to the URI encoder/decoder and not specific to using it for bodies here. If it doesn't work, it'd affect query items as well, which can commonly be arrays.

File: Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift [hunk 1]
```diff
@@ -177,12 +177,50 @@ extension Converter {
...
 
+    func convertURLEncodedFormToCodable<T: Decodable>(
+        _ data: Data
+    ) throws -> T {
+
+        let decoder = URIDecoder(
+            configuration: .init(
+                style: .form,
+                explode: true,
+                spaceEscapingCharacter: .percentEncoded,
```
Comment (czechboy0) [inline Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift:188]:
```suggestion
                spaceEscapingCharacter: .plus,
```

According to https://www.rfc-editor.org/rfc/rfc1866.html#section-8.2.1

> 1. The form field names and values are escaped: space
        characters are replaced by `+', and then reserved characters
        are escaped as per [URL];

File: Sources/OpenAPIRuntime/Errors/RuntimeError.swift [hunk 2]
```diff
@@ -23,6 +23,7 @@ internal enum RuntimeError: Error, CustomStringConvertible, LocalizedError, Pret
 
     // Data conversion
     case failedToDecodeStringConvertibleValue(type: String)
+    case failedToSerializeCodableData
```
Comment (czechboy0) [inline Sources/OpenAPIRuntime/Errors/RuntimeError.swift:26]: You can get rid of this, we'll never hit it with my suggestion above.

File: Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift [hunk 3]
```diff
@@ -177,12 +177,50 @@ extension Converter {
...
+                dateTranscoder: configuration.dateTranscoder
+            )
+        )
+        guard let uriString = String(data: data, encoding: .utf8) else {
```
Comment (czechboy0) [inline Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift:192]: Prefer `String(decoding: data, as: UTF8.self)` which isn't failable.

File: Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift [hunk 4]
```diff
@@ -177,12 +177,50 @@ extension Converter {
...
 
+    func convertBodyCodableToURLFormData<T: Encodable>(
+        _ value: T
+    ) throws -> Data {
+
+        let encoder = URIEncoder(
+            configuration: .init(
+                style: .form,
+                explode: true,
+                spaceEscapingCharacter: .percentEncoded,
```
Comment (czechboy0) [inline Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift:213]:
```suggestion
                spaceEscapingCharacter: .plus,
```

See above.

File: Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift [hunk 5]
```diff
@@ -177,12 +177,50 @@ extension Converter {
...
+                dateTranscoder: configuration.dateTranscoder
+            )
+        )
+        guard let data = try encoder.encode(value, forKey: "").data(using: .utf8) else {
```
Comment (czechboy0) [inline Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift:217]: To convert the string to Data, use `Data(string.utf8)` instead, as that's not failable and we don't need to handle the error.

File: Tests/OpenAPIRuntimeTests/Test_Runtime.swift [hunk 6]
```diff
@@ -98,6 +102,10 @@ class Test_Runtime: XCTestCase {
         """#
     }
 
+    var testStructURLFormString: String {
+        "age=3&name=Fluffz&type=Dog"
```
Comment (czechboy0) [inline Tests/OpenAPIRuntimeTests/Test_Runtime.swift:106]: Please make one of the strings have a space, and a special character, to verify escaping.

File: Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Client.swift [hunk 7]
```diff
@@ -234,6 +234,59 @@ final class Test_ClientConverterExtensions: Test_Runtime {
...
+        XCTAssertEqual(String(data: body, encoding: .utf8), testStructURLFormString)
+        XCTAssertEqual(
+            headerFields,
+            [
+                .init(name: "content-type", value: "application/x-www-form-urlencoded")
+            ]
+        )
+    }
+
+    func test_setRequiredRequestBodyAsURLEncodedForm_codable() throws {
```
Comment (czechboy0) [inline Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Client.swift:273]: I don't think you need this test, it's testing the same logic as above just comparing using data vs string, am I reading it correctly?

File: Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Client.swift [hunk 8]
```diff
@@ -234,6 +234,59 @@ final class Test_ClientConverterExtensions: Test_Runtime {
...
+    //    | client | set | request body | urlEncodedForm | codable | required | setRequiredRequestBodyAsURLEncodedForm |
+    func test_setRequiredRequestBodyAsURLEncodedForm_codable_string() throws {
+        var headerFields: [HeaderField] = []
+        let body = try converter.setRequiredRequestBodyAsURLEncodedForm(
+            testStructDetailed,
+            headerFields: &headerFields,
+            contentType: "application/x-www-form-urlencoded"
+        )
+
+        XCTAssertEqual(String(data: body, encoding: .utf8), testStructURLFormString)
```
Comment (czechboy0) [inline Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Client.swift:264]: It's fine to just compare it using `XCTAssertEqual(body, testStructURLFormData)`.

Alternatively, you could bring over `XCTAssertEqualStringifiedData` from the generator repo.

**Review by czechboy0 — CHANGES_REQUESTED** [review]
Looks great, just a few minor suggestions, otherwise almost ready to go.

File: Sources/OpenAPIRuntime/Conversion/Converter+Server.swift [hunk 9]
```diff
@@ -273,6 +273,34 @@ extension Converter {
         )
     }
 
+    //    | server | get | request body | URLEncodedForm | codable | optional | getOptionalRequestBodyAsURLEncodedForm |
```
[the remark this answers is not in the document]
Comment (bfrearson) [reply Sources/OpenAPIRuntime/Conversion/Converter+Server.swift:276]: Done, in the corresponding [generator PR](https://github.com/apple/swift-openapi-generator/pull/283)

File: Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Client.swift (hunk 7 above)
Comment (bfrearson) [reply Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Client.swift:273]: yeah, this was just for clarity when debugging a failed test. Importing the `XCTAssertEqualStringifiedData` means we only need one test now

**Comment by bfrearson** [comment]
@czechboy0 All changes complete on both generator & runtime 👍🏻

File: Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift [hunk 10]
```diff
@@ -177,12 +177,45 @@ extension Converter {
         try decoder.decode(T.self, from: data)
     }
 
+    func convertURLEncodedFormToCodable<T: Decodable>(
+        _ data: Data
+    ) throws -> T {
+
```
Comment (czechboy0) [inline Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift:183]:
```suggestion
```

File: Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift [hunk 11]
```diff
@@ -177,12 +177,45 @@ extension Converter {
...
     func convertBodyCodableToJSON<T: Encodable>(
         _ value: T
     ) throws -> Data {
         try encoder.encode(value)
     }
 
+    func convertBodyCodableToURLFormData<T: Encodable>(
+        _ value: T
+    ) throws -> Data {
+
```
Comment (czechboy0) [inline Sources/OpenAPIRuntime/Conversion/CurrencyExtensions.swift:205]:
```suggestion
```

File: Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Client.swift [hunk 12]
```diff
@@ -13,6 +13,7 @@
 //===----------------------------------------------------------------------===//
 import XCTest
 @_spi(Generated)@testable import OpenAPIRuntime
+import PetstoreConsumerTestCore
```
Comment (czechboy0) [inline Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Client.swift:16]:
```suggestion
```

File: Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Server.swift [hunk 13]
```diff
@@ -339,6 +339,26 @@ final class Test_ServerConverterExtensions: Test_Runtime {
         XCTAssertEqual(body, testStruct)
     }
 
+    //    | server | get | request body | URLEncodedForm | optional | getOptionalRequestBodyAsJSON |
```
Comment (czechboy0) [inline Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Server.swift:342]:
```suggestion
    //    | server | get | request body | urlEncodedForm | optional | getOptionalRequestBodyAsURLEncodedForm |
```

File: Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Server.swift [hunk 14]
```diff
@@ -339,6 +339,26 @@ final class Test_ServerConverterExtensions: Test_Runtime {
...
+    func test_getOptionalRequestBodyAsURLEncodedForm_codable() throws {
+        let body: TestPetDetailed? = try converter.getOptionalRequestBodyAsURLEncodedForm(
+            TestPetDetailed.self,
+            from: testStructURLFormData,
+            transforming: { $0 }
+        )
+        XCTAssertEqual(body, testStructDetailed)
+    }
+
+    //    | server | get | request body | URLEncodedForm | required | getRequiredRequestBodyAsJSON |
```
Comment (czechboy0) [inline Tests/OpenAPIRuntimeTests/Conversion/Test_Converter+Server.swift:352]:
```suggestion
    //    | server | get | request body | urlEncodedForm | required | getRequiredRequestBodyAsURLEncodedForm |
```

**Pull request merged**